src/ray/raylet/node_manager.cc

```diff
@@ -736,6 +741,73 @@ void NodeManager::DispatchTasks(
 }
 
+std::pair<std::shared_ptr<Worker>, bool> NodeManager::GetWorker(
+    const WorkerID &worker_id) {
+  bool is_worker = true;
+  auto worker = worker_pool_.GetRegisteredWorker(worker_id);
+  if (!worker) {
+    is_worker = false;
+    worker = worker_pool_.GetRegisteredDriver(worker_id);
+    if (!worker) {
+      return std::make_pair(nullptr, true);
+    }
+  }
+  return std::make_pair(worker, is_worker);
+}
+
+#define PREPROCESS_REQUEST(REQUEST_TYPE)                                            \
+  WorkerID worker_id = WorkerID::FromBinary(request.worker_id());                   \
+  do {                                                                              \
+    if (!PreprocessRequest<rpc::REQUEST_TYPE>(worker_id, #REQUEST_TYPE, request)) { \
+      return;                                                                       \
+    }                                                                               \
+  } while (0)
+
+template <typename Request>
+bool NodeManager::PreprocessRequest(const WorkerID &worker_id,
+                                    const std::string &request_name,
+                                    const Request &request) {
+  std::ostringstream string_stream;
+  string_stream << "Received a " << request_name << " request. Worker id "
+                << worker_id.Hex() << ".";
+  // NOTE(Joey Jiang): We check heartbeat actively because the heartbeats' callback won't
+  // be invoked until all the post requests in the grpc server have been finished.
+  // The raylet might be marked as dead due to missing heartbeats when too many requests
+  // are sent to this raylet.
+  int64_t expiry_delay =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::steady_clock::now() - heartbeat_timer_.expires_at())
+          .count();
+  int64_t expiry_delay_limit = RayConfig::instance().heartbeat_timeout_milliseconds() *
+                               RayConfig::instance().num_heartbeats_timeout() / 2;
+  // Timer's callback hasn't been called on the desired time.
+  if (expiry_delay > expiry_delay_limit) {
+    heartbeat_timer_.cancel();
+    Heartbeat();
+  }
+
+  auto rt = GetWorker(worker_id);
+  auto worker = rt.first;
+  // Worker process has been killed, we should discard this request.
+  if (!worker) {
+    string_stream << " Worker is not found in worker pool, request will be discarded.";
+    RAY_LOG(WARNING) << string_stream.str();
+    return false;
+  }
+  string_stream << " Is worker: " << (rt.second ? "true" : "false") << ". Worker pid "
+                << std::to_string(worker->Pid()) << ".";
+
+  // The worker process is being killing, we should discard this request.
+  if (worker->IsBeingKilled()) {
+    string_stream << " Worker process is being killed, request will be discarded.";
+    RAY_LOG(INFO) << string_stream.str();
+    return false;
+  }
+
+  RAY_LOG(DEBUG) << string_stream.str();
```

Comment:
Could we just use RAY_LOG(DEBUG) for the debug messages, instead of `straing_stream`? Thus we don't need to allocate these debug strings on production (where debug log is disabled).